SzczepionkaAPI/src/main/java/com/szczepionka/service/EmailService.java

```diff
@@ -0,0 +1,117 @@
+import org.springframework.stereotype.Service;
+import org.springframework.web.util.UriComponents;
+import org.springframework.web.util.UriComponentsBuilder;
+
+import java.io.IOException;
+import java.util.Date;
+import java.util.Optional;
+import java.util.Properties;
+
+@Service
+public class EmailService {
+
+    private final PatientRepository patientRepository;
+
+    @Value("${gmail.username}")
+    private String username;
+    @Value("${gmail.password}")
+    private String password;
+
+    public EmailService(PatientRepository patientRepository) {
+        this.patientRepository = patientRepository;
+    }
+
+
+    public PatientDTO sendmail(Long patientId) throws  AddressException, MessagingException, IOException {
+        Optional <Patient> patient = patientRepository.findById(patientId);
+        if(patient.isPresent()) {
+            PatientDTO patientDTO = PatientDTO.builder().
+                    email(patient.get().getEmail())
+                    .pesel(patient.get().getPesel())
+                    .postalCode(patient.get().getPostalCode())
+                    .referralId(patient.get().getReferralId())
+                    .build();
+
+            Properties props = new Properties();
+            props.put("mail.smtp.auth", "true");
+            props.put("mail.smtp.starttls.enable", "true");
+            props.put("mail.smtp.host", "smtp.gmail.com");
+            props.put("mail.smtp.port", "587");
+
+            Session session = Session.getInstance(props,
+                    new javax.mail.Authenticator() {
+                        protected PasswordAuthentication getPasswordAuthentication() {
+                            return new PasswordAuthentication(username, password);
+                        }
+                    });
+
+            Message msg = new MimeMessage(session);
+            msg.setFrom(new InternetAddress(username, false));
+
+            msg.setRecipients(Message.RecipientType.TO, InternetAddress.parse(patientDTO.getEmail()));
+            msg.setSubject("Potwierdzenie zapisu na szczepienie");
+            msg.setSentDate(new Date());
+
+            MimeBodyPart messageBodyPart = new MimeBodyPart();
+            messageBodyPart.setContent("Zostałeś/aś zrejstrowana na szczepienie "
+                    +"\nDane:\n"
+                    +"Pesel:\t"+maskString(patientDTO.getPesel(),3, 7, '*')
+                    +"\nKod pocztowy:\t" +patientDTO.getPostalCode()
+                    +"\nAby zobaczyć szczegóły szczepienia, kliknij w link poniżej\t"+ constructUri(patientDTO.getReferralId()),
+                    "text/plain; charset=utf-8");
+
+            Multipart multipart = new MimeMultipart();
+            multipart.addBodyPart(messageBodyPart);
+            msg.setContent(multipart);
+
+            Transport.send(msg);
+
+        return patientDTO;
+        }
+        return null;
+    }
+
+    private String maskString(String strText, int start, int end, char maskChar){
```

Comment:
I think that start, end and maskChar params are not necessary here. We can set this three values inside of method - there is no need to get it from another method. MaskString says everything - for me only String parameter will be fine. (It let us delete this code to check if start or end params are correct - I think it's overhead. Another thing is that name 'strText' have nothing in common with the thing that we are going to do. It's just pesel so this string can be named in that way. (we don't have nothing more to mask at the moment, but you can name it as sensitivePersonalInformation aswell).

Reply:
Maybe instead of this super fency method just use sth like:
`patientDTO.getPesel().replace(patientDTO.getPesel().substring(3, 7), "****"))`

Reply:
Method has been replaced for xNIGHTHAWKx's suggestion